### ISSUE TITLE
Leaky abstraction: loss returning Dual instead of TrackedReal

### ISSUE THREAD

**Issue by cscherrer** [FluxML/Flux.jl#366]
I'm playing with a maximum likelihood problem and have this setup:
```julia
using Flux
using Distributions

struct par
    μ
    logσ
end

Flux.@treelike par

data =  [-27.020,3.570,8.191,9.898,9.603,9.945,10.056]

p = par(param([0.0]),param(rand(7)))
```
Loss is just the negative log-likelihood:
```julia
loss(x) = -sum(logpdf.(Normal.(p.μ,exp.(p.logσ)),x))
typeof(loss(data))
```

This works find, and returns a `Flux.Tracker.TrackedReal{Float64}`. But then I try to make it more easily readable...
```julia
function loss(x)
    μ = p.μ
    σ = exp.(p.logσ)
    dists = Normal.(μ,σ)
    
    -sum(logpdf.(dists, x))
end

typeof(loss(data))
```
I would expect this to be equivalent, but it's now returning a `ForwardDiff.Dual{Nothing,Float64,2}`

**Comment by cscherrer** [comment]
@johnnychen94 That's pretty vague, can you please elaborate?

**Comment by MikeInnes** [comment]
This is behaving differently now -- it returns a `Float64` -- but it's still broken. However it works if I write

```julia
julia> function loss(x)
           μ = p.μ
           σ = exp.(p.logσ)
           -sum(logpdf.(Normal.(μ,σ), x))
       end
```

The issue is that our current implementation of `broadcast` can't handle anything that isn't scalar (for example, the `Normal` struct), so it just assume that's a non-differentiable object. OTOH if you fuse the `Normal` construction and the `logpdf` then the only inputs and outputs are real numbers again.

Not ideal but it'll be fixed with #669.
